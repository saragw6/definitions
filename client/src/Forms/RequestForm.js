import React, {Component} from 'react';
import Form from './FormComponent';
import {theme, ThemeProvider} from "../Libraries/ReactToolboxLibrary";

export default class RequestForm extends Component {
  constructor(props) {
    super(props);
    this.state = {Term:'', showErrorTerm: false, stateBar: false, snackbarMessage: ''};
  }

  createContent = () => {
    return (<div>
      <p>I'm collecting informal definitions to help people understand LGBTQ+ terms they don't know!</p>
      <p>You can request a term here that you want someone else to define.</p>
      <p> Check out the <a href='http://queerundefined.com/about'>about</a> page to read more about this project.</p>
    </div>)
  }

  handleSubmit(e) {
    if(e) e.preventDefault();
    const term = this.state.Term;
    if(term.trim().length === 0) {
      const snackBarMessage = 'You should provide a term!'
      this.setState({stateBar: true, showErrorTerm: true, snackbarMessage: snackBarMessage})
    }
    else {
      fetch('/requested/' + term, {method: 'POST'})
        .then(response => {
          let snackBarMessage = 'Your request for '.concat(term).concat(' was successful')
          if(response.status !== 200)
            snackBarMessage = 'Oh no! Something went wrong, please try again.'

          this.setState({stateBar: true, Term: '', snackbarMessage: snackBarMessage})
        });
    } }

  handleSnackbarClick = (event, instance) => {
    this.setState({ stateBar: false, snackbarMessage: ''});
  };

  handleSnackbarTimeout = (event, instance) => {
    this.setState({stateBar: false, snackbarMessage: ''});
  };

  handleChange = (label, isRequired, showError, value) => {
    if(value.length === 0 && isRequired)
      this.setState({...this.state, [label]: value, [showError]: true});
    else
      this.setState({...this.state, [label]: value, [showError]: false});
  };

  render() {
    const { showErrorTerm, Term, stateBar, snackbarMessage } = this.state;

    return (
      <ThemeProvider theme={theme}>
        <Form title='Request a Term'
              content={this.createContent()}
              inputs={[
                {labelInput: 'What term do you want defined?', isRequired: true, label:'Term', value: Term,
                  onChange:this.handleChange, showError: showErrorTerm}
                ]}
              onSubmit={this.handleSubmit.bind(this)}
              onClickSnackBar={this.handleSnackbarClick.bind(this)}
              onTimeoutSnackBar={this.handleSnackbarTimeout.bind(this)}
              stateBar={stateBar}
              snackbarMessage={snackbarMessage}
        >
        </Form>
      </ThemeProvider>
    )
  }
}
