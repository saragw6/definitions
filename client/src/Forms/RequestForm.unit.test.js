import React from 'react';
import renderer from "react-test-renderer";
import RequestForm from './RequestForm';

test('Request Form?', () => {
  const component = renderer.create(<RequestForm />);
  const jsonForm = component.toJSON();
  expect(jsonForm).toMatchSnapshot();
});