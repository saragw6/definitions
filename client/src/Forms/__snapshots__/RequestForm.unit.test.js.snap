exports[`test Request Form? 1`] = `
<div
  className="flex-container"
  data-testid="form-container">
  <form
    data-testid="request-form"
    onSubmit={[Function]}>
    <div
      className="form request">
      <div
        className="page-title">
        Request a Term
      </div>
      <div>
        <p>
          I\'m collecting informal definitions to help people understand LGBTQ+ terms they don\'t know!
        </p>
        <p>
          You can request a term here that you want someone else to define.
        </p>
        <p>
           Check out the 
          <a
            href="http://queerundefined.com/about">
            about
          </a>
           page to read more about this project.
        </p>
      </div>
      <span
        className="error">
        * Required
      </span>
      <div
        className="inputsBox">
        <div
          className="formInput">
          <label>
            What term do you want defined?
          </label>
          Â 
          <span
            className="error">
            *
          </span>
          <div
            className="lFVgC queerInput"
            data-react-toolbox="input">
            <input
              className="_4bZUj"
              data-testid="request-input"
              defaultValue={undefined}
              disabled={false}
              maxLength={50}
              name={undefined}
              onBlur={undefined}
              onChange={[Function]}
              onKeyPress={undefined}
              required={false}
              role="input"
              style={
                Object {
                  "background": "transparent",
                }
              }
              type="Text"
              value="" />
            <span
              className="_3FySS" />
            <span
              className="bMyi_"
              hidden={undefined}>
              Your answer
            </span>
            <span
              className="_1oTuT">
              0
              /
              50
            </span>
          </div>
        </div>
      </div>
      <div
        className="submitBox">
        <button
          className="_2Agdx _221ic _2Agdx _2GH_L _1ZxqC _3tTAW queerButton"
          data-react-toolbox="button"
          disabled={false}
          href={undefined}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchStart={[Function]}
          type="button">
          Submit
        </button>
      </div>
    </div>
  </form>
</div>
`;
